File: craft/commands/nodes.py
import click

from craft.client import get
from craft.output import print_item, print_table


@click.command("list")
def node_list():
    """List available infrastructure nodes.

    \b
    Shows node name, location, and status.
    Use 'craft node hardware <NODE_ID>' for detailed specs.
    """
    data = get("/nodes")
    items = data.get("data", data)
    if isinstance(items, dict):
        items = items.get("items", items.get("nodes", []))
    if isinstance(items, list) and items:
        rows = []
        for n in items:
            rows.append([
                n.get("id", ""),
                n.get("name", n.get("hostname", "")),
                n.get("location", ""),
                n.get("country", ""),
                n.get("status", ""),
            ])
        print_table(rows, ["ID", "Name", "Location", "Country", "Status"])
    else:
        print_item(data)


@click.command("hardware")
@click.argument("node_id", required=False, default=None)
def node_hardware(node_id):
    """Get node hardware info (CPU, RAM, storage).

    \b
    If NODE_ID is omitted, you can select interactively.
    """
    if not node_id:
        from craft.interactive import select_node
        node_id = select_node("Select node for hardware info")
    data = get(f"/nodes/{node_id}/hardware")
    print_item(data)

File: cmd/nodes.go
package cmd

import (
	"fmt"

	"github.com/CraftIntertech/craft-cli/internal/client"
	"github.com/CraftIntertech/craft-cli/internal/output"
	"github.com/spf13/cobra"
)

var nodeListCmd = &cobra.Command{
	Use:   "list",
	Short: "List all nodes",
	Run: func(cmd *cobra.Command, args []string) {
		result := client.Get("/nodes", nil, true)
		nodes := extractList(result, "nodes")

		var rows [][]string
		for _, n := range nodes {
			rows = append(rows, []string{
				getString(n, "id"),
				getString(n, "name"),
				getString(n, "location"),
				getString(n, "country"),
				getString(n, "status"),
			})
		}

		output.PrintTable(rows, []string{"ID", "Name", "Location", "Country", "Status"})
	},
}

var nodeHardwareCmd = &cobra.Command{
	Use:   "hardware [NODE_ID]",
	Short: "Show hardware details for a node",
	Run: func(cmd *cobra.Command, args []string) {
		nodeID := ""
		if len(args) > 0 {
			nodeID = args[0]
		}
		if nodeID == "" {
			nodeID = SelectNode("Select node for hardware info")
		}

		result := client.Get(fmt.Sprintf("/nodes/%s/hardware", nodeID), nil, true)
		output.PrintItem(extractObj(result))
	},
}

func init() {
	NodeCmd.AddCommand(nodeListCmd)
	NodeCmd.AddCommand(nodeHardwareCmd)
}
